### PR TITLE
propagate .meta info when replacing subgraphs in fx

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/torchdynamo/issues/1708

Our FX subgraph partitioner works by taking all of the original output nodes from a subgraph, and replacing it with a new `call_module` node in the graph.

If the original subgraph outputs had fake tensors and other metadata stored in their `.meta` attribute though, then this information was getting lost when we spliced in the subgraph.

Losing metadata on an FX graph also seems like an easy trap to fall into, so I'm wondering if there are any better guardrails that we can add. I ended up fixing in this PR by adding an optional kwarg to propagate meta info directly in the `fx.Node.replace_all_uses_with`, just because propagating metadata seems like a pretty core thing.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88198
* #88230
* #88197
* #87647
* #87646
* __->__ #87255

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
seems like a good change, but i wonder how we can tell if we're still missing cases, should the default be true?

**Comment by bdhirsh** [comment]
Yeah - I was pretty torn about whether or not copying metadata should be explicit vs. implicit.

Right now it's explicit - you specify `copy_metadata=True`, and you're on the hook for promising that the "new" node didn't have any metadata set on it - otherwise we error.

I wasn't sure about making it implicit, because `a.replace_all_uses_with(b)` seems like a general enough API that you might not actually want that - or that you might already have metadata on `b`, that you didn't want to stomp on.

I guess I could flip it so that we copy metadata by default, but **only** if there's no metadata field set on the new node (and if there is, maybe we should just silently not copy metadata instead of erroring). I'm open to either / other opinions though.

File: torch/fx/node.py [hunk 1]
```diff
@@ -468,7 +468,9 @@ def format_node(self,
     @compatibility(is_backward_compatible=True)
     def replace_all_uses_with(self,
```
Comment (bdhirsh) [inline torch/fx/node.py:469]: this is an FX public API change (it shouldn't be BC breaking though) - so I'll have to update the tests.

Who else should review public FX API changes? @SherlockNoMad ?

Comment (SherlockNoMad) [reply torch/fx/node.py:469]: This is not a BC breaking change. So it should be fine.

**Review by SherlockNoMad — APPROVED** [review]

**Pull request closed without merging**